react/lib/modal/Modal.tsx

```diff
@@ -1,70 +1,76 @@
-		</div>
-		{!slotContext.state.slotFooter ? null : (
-			<div className="modal-footer">
-				<Slot slotContent={slotContext.state.slotFooter} props={slotContext} />
+function DefaultSlotStructure<Data>(slotContext: ModalContext<Data>) {
+	return (
+		<>
+			{!slotContext.state.slotTitle ? null : (
+				<div className="modal-header">
+					<Slot slotContent={slotContext.state.slotHeader} props={slotContext}></Slot>
+				</div>
+			)}
+			<div className="modal-body">
+				<Slot slotContent={slotContext.state.slotDefault} props={slotContext} />
 			</div>
-		)}
-	</>
-);
+			{!slotContext.state.slotFooter ? null : (
+				<div className="modal-footer">
+					<Slot slotContent={slotContext.state.slotFooter} props={slotContext} />
+				</div>
+			)}
+		</>
+	);
+}
 
-const defaultConfig: Partial<ModalProps> = {
+const defaultConfig: Partial<ModalProps<any>> = {
 	slotHeader: DefaultSlotHeader,
 	slotStructure: DefaultSlotStructure,
 };
 
-export const Modal = forwardRef(function Modal(props: PropsWithChildren<Partial<ModalProps>>, ref) {
-	const [state, widget] = useWidgetWithConfig(createModal, props, 'modal', {...defaultConfig, slotDefault: props.children});
+function ModalInner<Data>(props: PropsWithChildren<Partial<ModalProps<Data>>>, ref: Ref<unknown> | undefined) {
+	const [state, widget] = useWidgetWithConfig(createModal<Data>, props, 'modal', {...defaultConfig, slotDefault: props.children});
 	useImperativeHandle(ref, () => widget.api, []);
 	const refSetBackdrop = useDirective(widget.directives.backdropDirective);
 	const refSetModal = useDirective(widget.directives.modalDirective);
-	const slotContext: ModalContext = {
+	const slotContext: ModalContext<Data> = {
 		state,
 		widget: toSlotContextWidget(widget),
 	};
 	return (
-		<>
-			<Portal container={state.container}>
-				{state.backdropHidden ? null : <div className={`modal-backdrop ${state.backdropClass}`} ref={refSetBackdrop} />}
-				{state.hidden ? null : (
-					<div className={`modal d-block ${state.className}`} ref={refSetModal} onClick={widget.actions.modalClick}>
-						<div className="modal-dialog">
-							<div className="modal-content">
-								<Slot slotContent={state.slotStructure} props={slotContext} />
-							</div>
+		<Portal container={state.container}>
+			{state.backdropHidden ? null : <div className={`modal-backdrop ${state.backdropClass}`} ref={refSetBackdrop} />}
+			{state.hidden ? null : (
+				<div className={`modal d-block ${state.className}`} ref={refSetModal} onClick={widget.actions.modalClick}>
+					<div className="modal-dialog">
+						<div className="modal-content">
+							<Slot slotContent={state.slotStructure} props={slotContext} />
 						</div>
 					</div>
-				)}
-			</Portal>
-		</>
+				</div>
+			)}
+		</Portal>
 	);
-});
+}
+
+export const Modal = forwardRef(ModalInner) as <Data>(
+	props: PropsWithChildren<Partial<ModalProps<Data>>> & {ref?: Ref<unknown> | undefined}
```

Comment:
`Ref<unknown>` is suspicious to me. Why don't you use the correct type `Ref<ModalWidget<Data>['api']>`?

Reply:
Thx for the help, it is not correctly typed.